lib/bandit/http1/handler.ex

```diff
@@ -88,6 +100,34 @@ defmodule Bandit.HTTP1.Handler do
     end
   end
 
+  defp maybe_upgrade_h2c(state, req, transport_info, method, request_target, headers) do
+    with {:http_2_enabled, true} <- {:http_2_enabled, state.http_2_enabled},
+         {:upgrade, "h2c"} <- {:upgrade, Bandit.Headers.get_header(headers, "upgrade")},
+         %Bandit.TransportInfo{secure?: false} <- transport_info,
+         {:ok, connection_headers} <- Bandit.Headers.get_connection_header_keys(headers),
+         {:settings, [{"http2-settings", settings_payload}]} <-
+           {:settings, Enum.filter(headers, fn {key, _value} -> key == "http2-settings" end)},
+         {:ok, settings_payload} <- Base.url_decode64(settings_payload),
+         resp_headers = [{"connection", "Upgrade"}, {"upgrade", "h2c"}],
+         {:ok, _sent_body, _req} <- Bandit.HTTP1.Adapter.send_resp(req, 101, resp_headers, <<>>) do
+      headers =
+        Enum.reject(headers, fn {key, _value} ->
+          key == "connection" || key in connection_headers
+        end)
```

Comment:
I'm not sure we need to be removing the connection headers (I don't see anything in RFC7540§3.2 about it)?

Reply:
I did this because of the following specific paragraph from RFC7540§3.2.1

> Since the upgrade is only intended to apply to the immediate
   connection, a client sending the HTTP2-Settings header field MUST
   also send "HTTP2-Settings" as a connection option in the Connection
   header field to prevent it from being forwarded (see Section 6.1 of
   [RFC7230]).

But that is regarding proxy servers and gateways (as per RFC7230§6.1) so probably we can keep them. Just let me know which one it is.

Reply:
Also worth mentioning: RFC9113§8.2.2

Reply:
Yeah, we should be keeping them. Whether or not those headers get forwarded (or what gets done with them in any way, really) is the Plug's call, not ours.

Reply:
But if we don't remove the "connection" header here and pass it through, then [stream_task.ex:151](https://github.com/mtrudel/bandit/blob/main/lib/bandit/http2/stream_task.ex#L151) will complain!

Reply:
Ah! Good point! I'll have a closer look at this tomorrow morning (~14:00 UTC)

Reply:
Yep, your original implementation is correct; we should be stripping those.